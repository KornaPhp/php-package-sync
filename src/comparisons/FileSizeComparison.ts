import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';

export class FileSizeComparison extends Comparison {
    protected filesize1 = 0;
    protected filesize2 = 0;

    public compare(requiredScore: number | null): Comparison {
        this.filesize1 = this.file?.sizeOnDisk ?? 0;
        this.filesize2 = this.repoFile?.sizeOnDisk ?? 0;

        this.score = this.percentage();
        this.comparisonPassed = !this.meetsRequirement(requiredScore);

        if (!this.passed()) {
            this.createIssue(null);
        }

        return this;
    }

    public getKind(): ComparisonKind {
        return ComparisonKind.ALLOWED_SIZE_DIFFERENCE_EXCEEDED;
    }

    public meetsRequirement(percentage): boolean {
        if (percentage === null) {
            percentage = this.file?.requiredScores?.size ?? 15;
        }

        const pctDecimal = percentage * 0.01;

        return this.difference() > Math.round(this.filesize1 * pctDecimal);
    }

    public prettyScore(): string {
        if (this.score === null) {
            return '';
        }

        return (this.score?.toFixed(2) ?? '0.0') + '%';
    }

    public difference() {
        return this.filesize1 - this.filesize2;
    }

    public percentage(): number {
        if (this.difference() === 0) {
            return 0;
        }

        return (this.difference() / this.filesize1) * 100;
    }
}
