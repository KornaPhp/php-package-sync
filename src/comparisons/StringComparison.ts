import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';

const { compareTwoStrings } = require('string-similarity');

export class StringComparison extends Comparison {
    public compare(requiredScore: number | null): Comparison {
        this.score = compareTwoStrings(this.file?.contents ?? '', this.repoFile?.contents ?? '');
        this.comparisonPassed = this.meetsRequirement(requiredScore ?? 0);

        if (!this.passed()) {
            this.createIssue(null);
        }

        return this;
    }

    public getKind(): ComparisonKind {
        return ComparisonKind.FILE_NOT_SIMILAR_ENOUGH;
    }

    public meetsRequirement(percentage) {
        if (percentage === null || percentage === 0) {
            percentage = this.file?.requiredScores?.similar ?? 15;
        }

        return super.meetsRequirement(percentage);
    }
}
