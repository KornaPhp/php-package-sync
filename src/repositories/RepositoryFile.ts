/* eslint-disable no-unused-vars */

import { ComparisonScoreRequirements } from '../types/ComparisonScoreRequirements';
import { File } from '../lib/File';

export class RepositoryFile extends File {
    public shouldIgnore = false;
    public shouldCompare = false;
    public requiredScores: ComparisonScoreRequirements = { similar: 0, size: 0 };

    /**
     *
     * @param repository Repository class
     * @param fn string
     * @param contents string | null
     */
    constructor(public repository: any, protected fn: string, contents: string | null = null) {
        super(fn, contents);
    }

    get relativeName() {
        return this.fn.replace(this.repository.path + '/', '');
    }

    public withOptions(ignore: boolean, compare: boolean, requiredScores: ComparisonScoreRequirements) {
        this.shouldIgnore = ignore;
        this.shouldCompare = compare;
        this.requiredScores = requiredScores;
        //
        return this;
    }

    public processTemplate(): string {
        return super.processTemplate(this.repository.name);
    }
}
