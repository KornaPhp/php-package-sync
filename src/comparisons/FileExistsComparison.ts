/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */

import { RepositoryFile } from '../repositories/RepositoryFile';
import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';

export class FileExistsComparison extends Comparison {
    protected kind: ComparisonKind = ComparisonKind.FILE_NOT_FOUND;

    public compare(requiredScore: number | null): Comparison {
        this.score = 0;
        this.comparisonPassed = this.meetsRequirement(requiredScore ?? 0);

        if (!this.comparisonPassed) {
            this.kind = this.file?.isFile() ?? false ? ComparisonKind.FILE_NOT_FOUND : ComparisonKind.DIRECTORY_NOT_FOUND;

            this.createIssue(null);
        }

        return this;
    }

    public getKind(): ComparisonKind {
        return this.kind;
    }

    public meetsRequirement(percentage): boolean {
        if (this.file === null) {
            return false;
        }

        return (this.file?.shouldIgnore ?? false) || this.repo.hasFile(<RepositoryFile>(<unknown>this.file));
    }
}
