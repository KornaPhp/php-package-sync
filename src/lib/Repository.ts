/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */

import { basename } from 'path';
import { app } from '../Application';
//import { PackageIssue } from '../issues/PackageIssue';
import { RepositoryIssue } from '../issues/RepositoryIssue';
import { Composer } from './composer/Composer';
import { getFileList, isDirectory } from './helpers';
import { RepositoryFile } from './RepositoryFile';

export enum RepositoryKind {
    SKELETON = 'skeleton',
    PACKAGE = 'package',
}

export class Repository {
    protected composerData: Composer;
    protected loadedFiles = false;
    protected fileList: RepositoryFile[] = [];
    public issues: RepositoryIssue[] = [];

    constructor(public path: string, public kind: RepositoryKind) {
        this.composerData = Composer.createFromPath(path);
    }

    static create(path: string, kind: RepositoryKind) {
        return new Repository(path, kind);
    }

    get name() {
        return basename(this.path);
    }

    get composer() {
        return this.composerData;
    }

    get packages() {
        return this.composer.packages('all');
    }

    get files() {
        if (!this.loadedFiles) {
            this.loadFiles();
        }

        return this.fileList;
    }

    getFile(relativeName: string): RepositoryFile | null {
        return this.files.find(f => f.relativeName === relativeName) || null;
    }

    hasFile(file: RepositoryFile): boolean {
        return this.files.find(f => f.relativeName === file.relativeName) !== undefined;
    }

    hasIssues() {
        return this.issues.length > 0;
    }

    isSkeleton() {
        return this.kind === RepositoryKind.SKELETON;
    }

    isPackage() {
        return this.kind === RepositoryKind.PACKAGE;
    }

    protected getFileList(directory: string, basePath: string | null = null): RepositoryFile[] {
        const result: RepositoryFile[] = [];

        getFileList(directory, basePath, true)
            .filter(fn => !app.shouldIgnoreFile(fn.replace(`${basePath}/`, '')))
            .forEach(fqName => {
                const rfile = new RepositoryFile(this, fqName);

                const relativeName = basePath ? fqName.replace(`${basePath}/`, '') : fqName;
                const isFile = !isDirectory(fqName);

                if (!result.find(item => item.relativeName === relativeName)) {
                    result.push(
                        rfile.withOptions(
                            app.shouldIgnoreFile(relativeName),
                            rfile.isFile() && app.shouldCompareFile(fqName),
                            app.getFileScoreRequirements(fqName),
                        ),
                    );
                    //     {
                    //     isFile: !isPath,
                    //     name: fqName,
                    //     relativeName: relativeName,
                    //     filesize: fileSize(fqName),
                    //     shouldIgnore: app.shouldIgnoreFile(relativeName),
                    //     shouldCompare: !isPath && app.shouldCompareFile(fqName),
                    //     requiredScores: app.getFileScoreRequirements(fqName),
                    // });

                    if (rfile.isDirectory()) {
                        result.push(...this.getFileList(fqName, basePath ?? directory));
                    }
                }
            });

        return result;
    }

    public loadFiles() {
        this.fileList = this.getFileList(this.path, this.path);
        this.loadedFiles = true;

        return this;
    }
}
