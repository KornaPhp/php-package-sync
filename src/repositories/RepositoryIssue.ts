/* eslint-disable no-unused-vars */

import { ComparisonKind } from '../types/FileComparisonResult';
import { Repository } from './Repository';
import { RepositoryFile } from './RepositoryFile';
//import Fixer from './fixers/Fixer';
import { classOf } from '../lib/helpers';

export class RepositoryIssue {
    protected _fixers: any[] = [];
    public resolvedByFixer = 'none';
    public resolvedNotes: string[] = [];
    public pending = false;

    constructor(
        public result: any,
        public name: string,
        public srcFile: RepositoryFile | null,
        public destFile: RepositoryFile | null,
        public skeleton: Repository,
        public repository: Repository,
        public resolved: boolean = false,
        public note: string | null = null,
        public context: any | null = null,
    ) {
        //
    }

    get fixers() {
        const same = (a, b, method) => (a[method]() && b[method]()) || (!a[method]() && !b[method]());
        const isTrue = (a, method) => (a[method]() ? 1 : -1);
        const methodCompare = (a, b, method) => (same(a, b, method) ? 0 : isTrue(a, method));

        return this._fixers.sort((a, b) => methodCompare(a, b, 'runsFixers'))
            .sort((a, b) => methodCompare(a, b, 'isRisky'));
    }

    get kind(): ComparisonKind {
        return this.result.kind;
    }

    get score(): number | string {
        if (this.result.score === 0) {
            return '-';
        }

        if (typeof this.result.score === 'number') {
            return this.result.score.toFixed(3);
        }

        return this.result.score;
    }

    get sourcefile(): RepositoryFile {
        return <RepositoryFile>this.srcFile;
    }

    get targetfile(): RepositoryFile {
        return <RepositoryFile>this.destFile;
    }

    public addFixer(fixer: any) {
        if (this.fixers.find(f => f.getName() === fixer.getName()) === undefined) {
            this.fixers.push(fixer);
        }

        return this;
    }

    public is(kind: ComparisonKind): boolean {
        return this.kind === kind;
    }

    public resolve(resolvedByFixer: string | any) {
        if (typeof resolvedByFixer !== 'string') {
            resolvedByFixer = classOf(resolvedByFixer)
                .prettyName();
        }

        this.resolvedByFixer = <string>resolvedByFixer;
        this.resolved = true;

        return this;
    }

    public addResolvedNote(note: string, prepend = false) {
        if (prepend) {
            this.resolvedNotes.unshift(note);
            return this;
        }

        this.resolvedNotes.push(note);

        return this;
    }

    public disableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.disable();
            }
        });

        return this;
    }

    public enableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.enable();
            }
        });

        return this;
    }
}
