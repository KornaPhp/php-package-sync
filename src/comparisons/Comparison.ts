/* eslint-disable no-unused-vars */

import { Repository } from '../repositories/Repository';
import { RepositoryFile } from '../repositories/RepositoryFile';
import { RepositoryIssue } from '../repositories/RepositoryIssue';
import { ComparisonKind } from '../types/FileComparisonResult';

export abstract class Comparison {
    protected comparisonPassed = false;
    protected _score: number | string | null = null;

    get score() {
        return this._score;
    }

    set score(value: number | string | null) {
        this._score = value;
    }

    constructor(
        public skeleton: Repository,
        public repo: Repository,
        public file: RepositoryFile | null,
        public repoFile: RepositoryFile | null,
    ) {
        //
    }

    static create<T extends Comparison>(
        this: new (...args: any[]) => T,
        skeleton: Repository,
        repo: Repository,
        file: RepositoryFile | null,
        repoFile: RepositoryFile | null,
    ): T {
        return new this(skeleton, repo, file, repoFile);
    }

    public abstract compare(requiredScore: number | null): Comparison;

    public abstract getKind(): ComparisonKind;

    public prettyScore(): string {
        if (this.score === null || this.score === 0) {
            return '-';
        }

        if (typeof this.score === 'string') {
            return this.score;
        }

        return this.score?.toFixed(3) ?? '';
    }

    protected createIssue(additional: Record<string, unknown> | null): void {
        const compareResult = {
            kind: this.getKind(),
            score: this.prettyScore(),
            ...(additional ?? {}),
        };

        this.repo.issues.push(
            new RepositoryIssue(
                compareResult,
                this.file?.relativeName ?? '.',
                this.file,
                this.repoFile,
                this.skeleton,
                this.repo,
                false,
            ),
        );
    }

    public passed(): boolean {
        return this.comparisonPassed;
    }

    public failed(): boolean {
        return !this.passed();
    }

    public meetsRequirement(minimum: number | null): boolean {
        return (this.score ?? 0) >= (minimum ?? 15);
    }
}
