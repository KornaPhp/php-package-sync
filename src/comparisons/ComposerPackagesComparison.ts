/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/ban-ts-comment */

import { ComparisonKind } from '../types/FileComparisonResult';
import { Comparison } from './Comparison';
import { Composer } from '../lib/composer/Composer';
import { RepositoryIssue } from '../repositories/RepositoryIssue';

const semver = require('semver');

export class ComposerPackagesComparison extends Comparison {
    protected kind: ComparisonKind = ComparisonKind.PACKAGE_NOT_USED;

    protected getVersionDiff(repoVersion: string, newVersion: string): string | null {
        const newPart = newVersion.split('|')
            .pop() ?? '0.0';
        const repoPart = repoVersion.split('|')
            .pop() ?? '0.0';

        const diff = semver.diff(semver.coerce(newPart), semver.coerce(repoPart));

        return ['major', 'minor', 'patch'].includes(diff) ? diff : null;
    }

    public compare(requiredScore: number | null): Comparison {
        this.score = 0;

        const skeletonComposer = Composer.createFromPath(this.skeleton.path);
        const repositoryComposer = Composer.createFromPath(this.repo.path);

        let issueCounter = 0;

        skeletonComposer.packages('all')
            .forEach(pkg => {
                if (!repositoryComposer.hasPackage(pkg.name, pkg.section)) {
                    this.kind = ComparisonKind.PACKAGE_NOT_USED;
                    this.score = 0;

                    this.createIssue({
                        name: pkg.name,
                        context: pkg.section,
                        skeletonPath: this.skeleton.path,
                        repositoryPath: this.repo.path,
                    });

                    issueCounter++;

                    return;
                }

                const repositoryPackage = repositoryComposer.package(pkg.name);
                const versionDiff = this.getVersionDiff(repositoryPackage.version, pkg.version);

                // treat version strings like '^8.1|^9.5' and '^9.5' as a non-issue
                if (repositoryPackage.version.includes(pkg.version)) {
                    return;
                }

                if (versionDiff !== null) {
                    this.kind = ComparisonKind.PACKAGE_VERSION_MISMATCH;
                    this.score = versionDiff;

                    this.createIssue({
                        name: pkg.name,
                        context: pkg,
                        skeletonPath: this.skeleton.path,
                        repositoryPath: this.repo.path,
                    });

                    issueCounter++;

                    return;
                }
            });

        this.comparisonPassed = issueCounter === 0;

        return this;
    }

    public getKind(): ComparisonKind {
        return this.kind;
    }

    public meetsRequirement(percentage): boolean {
        return true;
    }

    protected createIssue(additional: Record<string, unknown> | null): void {
        const compareResult = {
            kind: this.getKind(),
            score: this.prettyScore(),
            ...(additional ?? {}),
        };

        this.repo.issues.push(
            new RepositoryIssue(
                compareResult,
                // @ts-ignore
                compareResult.name,
                this.file,
                this.repoFile,
                this.skeleton,
                this.repo,
                false,
            ),
        );
    }
}
