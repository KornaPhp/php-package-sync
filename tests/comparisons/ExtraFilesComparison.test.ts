/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import { ExtraFilesComparison } from '../../src/comparisons/ExtraFilesComparison';
import { Repository, RepositoryKind } from '../../src/repositories/Repository';

let skeleton: Repository;

beforeEach(() => {
    skeleton = Repository.create(__dirname + '/../data/test-skeleton', RepositoryKind.SKELETON);
});

it('passes when there are no files in the repo that are not in the skeleton', () => {
    const repo = Repository.create(__dirname + '/../data/test-package', RepositoryKind.PACKAGE);

    const comparisonObj = ExtraFilesComparison.create(skeleton, repo, null, null);

    comparisonObj.compare(null);

    expect(comparisonObj.passed())
        .toBeTruthy();
});

it('fails when there are files that exist only in the repo directory', () => {
    const repo = Repository.create(__dirname + '/../data/test-package-2', RepositoryKind.PACKAGE);

    const comparisonObj = ExtraFilesComparison.create(skeleton, repo, null, null);

    comparisonObj.compare(null);

    expect(comparisonObj.passed())
        .toBeFalsy();
});

it('does not create issues when there are no files in the repo that are not in the skeleton', () => {
    const repo = Repository.create(__dirname + '/../data/test-package', RepositoryKind.PACKAGE);

    const comparisonObj = ExtraFilesComparison.create(skeleton, repo, null, null);

    comparisonObj.compare(null);

    expect(repo.issues)
        .toHaveLength(0);
});

it('creates issues when there are files that exist only in the repo directory', () => {
    const repo = Repository.create(__dirname + '/../data/test-package-2', RepositoryKind.PACKAGE);

    const comparisonObj = ExtraFilesComparison.create(skeleton, repo, null, null);

    comparisonObj.compare(null);

    const issues = repo.issues.map(issue => ({
        name: issue.name,
        kind: issue.result.kind,
    }));

    expect(repo.issues.length)
        .toBeGreaterThan(0);
    expect(issues)
        .toMatchSnapshot();
});
