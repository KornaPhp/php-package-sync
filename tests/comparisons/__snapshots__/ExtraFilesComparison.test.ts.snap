// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates issues when there are files that exist only in the repo directory 1`] = `
[
  {
    "kind": "extra_file",
    "name": "one.txt",
  },
  {
    "kind": "extra_file",
    "name": "two.txt",
  },
]
`;
